src/BoundsInference.cpp

```diff
@@ -109,27 +110,178 @@ class BoundsInference : public IRMutator {
+                    }
+                    values.push_back(CondValue(const_true(), cond_val));
+                } else {
+                    values.push_back(CondValue(const_true(), val));
+                }
+            }
+            if (is_update) {
+                for (const Expr &arg : def.args()) {
+                    if (!predicates.empty()) {
+                        Expr cond_arg = Call::make(arg.type(),
+                                                   Internal::Call::if_then_else,
+                                                   {likely(predicates[0]), arg, make_zero(arg.type())},
+                                                   Internal::Call::PureIntrinsic);
+                        for (size_t i = 1; i < predicates.size(); ++i) {
+                            cond_arg = Call::make(cond_arg.type(),
+                                                  Internal::Call::if_then_else,
+                                                  {likely(predicates[i]), cond_arg, make_zero(cond_arg.type())},
+                                                  Internal::Call::PureIntrinsic);
+                        }
+                        args.push_back(CondValue(const_true(), cond_arg));
+                    } else {
+                        args.push_back(CondValue(const_true(), arg));
+                    }
+                }
+            }
+
+            const vector<Specialization> &specializations = def.specializations();
+            for (size_t i = specializations.size(); i > 0; i--) {
+                Expr s_cond = specializations[i-1].condition;
+                const Definition &s_def = specializations[i-1].definition;
+
+                // Else case (i.e. specialization condition is false)
+                for (CondValue &i : args) {
+                    i.cond = simplify(!s_cond && i.cond);
+                }
+                for (CondValue &i : values) {
+                    i.cond = simplify(!s_cond && i.cond);
+                }
+
+                // Then case (i.e. specialization condition is true)
+                vector<CondValue> s_args, s_values;
+                std::tie(s_args, s_values) = compute_exprs_helper(s_def, is_update);
+                for (CondValue &i : s_args) {
+                    i.cond = simplify(s_cond && i.cond);
+                }
+                for (CondValue &i : s_values) {
+                    i.cond = simplify(s_cond && i.cond);
+                }
+                values.insert(values.end(), s_values.begin(), s_values.end());
+                args.insert(args.end(), s_args.begin(), s_args.end());
+            }
+
+            // Optimization: If the args/values across specializations including
+            // the default case, are the same, we can combine those args/values
+            // into one arg/value with a const_true() condition for the purpose
+            // of bounds inference.
+            if (values.size() > 1) {
```

Comment:
Perhaps it's possible to avoid duplicated code here like so:
    for (vector<CondValue> &vec : {args, values}) {
     ... the original code ...
    }